### DISCUSSION TITLE
New Collection expressions initializer allows null for non nullable Collections

### DISCUSSION THREAD

**Discussion by TheCell** [dotnet/csharplang#8339] [Q&A] [answered]
Is this a Visual Studio Problem, C# Problem or is that allowed behaviour?

This Code compiles and runs, resulting in a list with one entry that is null, even though the List accepts only non nullable strings.
``` csharp
string? testString = null;
List<string> testList = [testString];
```
![image](https://github.com/user-attachments/assets/611d6407-7818-4fdb-af97-72df574811f0)

This code does not compile:
``` csharp
List<string> testList3 = new List<string> { null };
List<string> testList4 = new List<string>();
testList4.Add(null);
```
![image](https://github.com/user-attachments/assets/988131dc-7f1e-48fc-8297-e249230f7de1)

We are using the Roslyn analyzer.
Our Solution is using the following:
```
<Nullable>enable</Nullable>
<WarningsAsErrors>nullable</WarningsAsErrors>
<EnableNETAnalyzers>true</EnableNETAnalyzers>
<AnalysisLevel>latest</AnalysisLevel>
```

I think that's a bug?

**Comment by HaloFour** [comment]
Both snippets of code compile successfully.  The C# compiler will issue warnings about potential nulls, but it is not a compiler error.  Are you using a third-party analyzer in Visual Studio, like R#?  That might be giving you the errors on the second code snippet.

**Reply by TheCell** [reply]
Using the Roslyn analyzer.

Our Solution is using the following:
```
<Nullable>enable</Nullable>
<WarningsAsErrors>nullable</WarningsAsErrors>
<EnableNETAnalyzers>true</EnableNETAnalyzers>
<AnalysisLevel>latest</AnalysisLevel>
```

**Comment by 333fred** [comment]
This looks like https://github.com/dotnet/roslyn/issues/71522, @RikkiGibson to confirm.

**Comment by dlebansais** [comment]
There are about five different ways to insert `null` values in your program without the compiler or analyzers noticing. I wouldn't call that a bug, but a limitation of what they can do to help you find the error before the code throws an exception at runtime.

**Comment by cston** [comment, marked as answer]
This is one of the remaining work items in https://github.com/dotnet/roslyn/issues/68786:

> - [ ] Checking conversion of elements to iteration type for *collection initializer types*
